apollo-router/src/plugins/authentication.rs

```diff
@@ -0,0 +1,901 @@
+>;
+
+pub(crate) const AUTHENTICATION_SPAN_NAME: &str = "authentication plugin";
+
+const DEFAULT_AUTHENTICATION_NETWORK_TIMEOUT: Duration = Duration::from_secs(15);
+
+const DEFAULT_AUTHENTICATION_COOLDOWN: Duration = Duration::from_secs(15);
+
+static COOLDOWN: Lazy<AtomicBool> = Lazy::new(|| AtomicBool::new(false));
+
+static CLIENT: Lazy<Result<Client, BoxError>> = Lazy::new(|| {
+    apollo_graph_reference().ok_or(LicenseError::MissingGraphReference)?;
+    Ok(Client::new())
+});
+
+struct AuthenticationPlugin {
+    configuration: JWTConf,
+    jwks: SharedDeduplicate,
+    jwks_url: Url,
+}
+
+#[derive(Clone, Debug, Default, Deserialize, JsonSchema)]
+struct JWTConf {
+    /// Retrieve our JWK Set from here
+    jwks_url: String,
+    /// HTTP header expected to contain JWT
+    #[serde(default = "default_header_name")]
+    header_name: String,
+    /// Header prefix
+    #[serde(default = "default_header_prefix")]
+    header_prefix: String,
+    /// JWKS retrieval cooldown
+    #[serde(deserialize_with = "humantime_serde::deserialize", default)]
+    #[schemars(with = "String", default)]
+    cooldown: Option<Duration>,
+}
+
+// This is temporary. It will be removed when the plugin is promoted
+// from experimental.
+#[derive(Clone, Debug, Default, Deserialize, JsonSchema)]
+struct ExperimentalConf {
+    /// The JWT configuration
+    jwt: JWTConf,
+}
+
+// We may support additional authentication mechanisms in future, so all
+// configuration (which is currently JWT specific) is isolated to the
+// JWTConf structure.
+/// Authentication
+#[derive(Clone, Debug, Default, Deserialize, JsonSchema)]
+struct Conf {
+    /// The experimental configuration
+    experimental: ExperimentalConf,
+}
+
+fn default_header_name() -> String {
+    http::header::AUTHORIZATION.to_string()
+}
+
+fn default_header_prefix() -> String {
+    "Bearer".to_string()
+}
+
+#[async_trait::async_trait]
+impl Plugin for AuthenticationPlugin {
+    type Config = Conf;
+
+    async fn new(init: PluginInit<Self::Config>) -> Result<Self, BoxError> {
+        let url: Url = Url::from_str(&init.config.experimental.jwt.jwks_url)?;
+        let getter: Box<dyn Fn(Url) -> DeduplicateFuture<JwkSet> + Send + Sync + 'static> =
+            Box::new(|url: Url| -> DeduplicateFuture<JwkSet> {
+                let fut = async {
+                    let data = if url.scheme() == "file" {
+                        #[cfg(not(test))]
+                        apollo_graph_reference()
+                            .ok_or(LicenseError::MissingGraphReference)
+                            .ok()?;
+                        let path = url.to_file_path().ok()?;
+                        read_to_string(path).await.ok()?
+                    } else {
+                        let my_client = CLIENT.as_ref().map_err(|e| e.to_string()).ok()?.clone();
+
+                        my_client
+                            .get(url)
+                            .header(ACCEPT, "application/json")
```

Comment:
We have a constant for this `"application/json"`

Reply:
done: 660f6552d824930ee265404b4d90b3221dcb6461